vsphere/resource_vsphere_compute_cluster.go

```diff
@@ -1455,6 +1505,60 @@ func expandVsanDatastoreConfig(d *schema.ResourceData, meta interface{}) (*vsant
 	return conf, nil
 }
 
+func buildVsanStretchedClusterReq(d *schema.ResourceData, cluster types.ManagedObjectReference) (*vsantypes.VSANVcConvertToStretchedCluster, error) {
+	log.Printf("[DEBUG] building vsan stretched cluster request...")
+	conf := d.Get("vsan_stretched_cluster_conf").(*schema.Set).List()[0].(map[string]interface{})
+
+	if conf["preferred_fault_domain_host_ids"] == "" {
+		return nil, fmt.Errorf("preferred_fault_domain_host_ids field is required in vsan_stretched_cluster_conf when enabling stretched cluster!")
+	}
+
+	if conf["secondary_fault_domain_host_ids"] == "" {
+		return nil, fmt.Errorf("secondary_fault_domain_host_ids field is required in vsan_stretched_cluster_conf when enabling stretched cluster!")
+	}
+
+	if conf["witness_node"] == "" {
+		return nil, fmt.Errorf("witness_node field is required in vsan_stretched_cluster_conf when enabling stretched cluster!")
+	}
+
+	witness_node := []string{}
+	witness_node = append(witness_node, conf["witness_node"].(string))
+	witness := structure.SliceStringsToManagedObjectReferences(witness_node, "HostSystem")
+
+	fault_domain_config := vsantypes.VimClusterVSANStretchedClusterFaultDomainConfig{
+		FirstFdName:   conf["preferred_fault_domain_name"].(string),
+		FirstFdHosts:  structure.SliceInterfacesToManagedObjectReferences(conf["preferred_fault_domain_host_ids"].(*schema.Set).List(), "HostSystem"),
+		SecondFdName:  conf["secondary_fault_domain_name"].(string),
+		SecondFdHosts: structure.SliceInterfacesToManagedObjectReferences(conf["secondary_fault_domain_host_ids"].(*schema.Set).List(), "HostSystem"),
+	}
+
+	// TODO: make diskmapping configurable.
+	return &vsantypes.VSANVcConvertToStretchedCluster{
+		This:              vsan.VsanVcStretchedClusterSystem,
+		Cluster:           cluster.Reference(),
+		FaultDomainConfig: fault_domain_config,
+		WitnessHost:       witness[0],
+		PreferredFd:       conf["preferred_fault_domain_name"].(string),
```

Comment:
Is this piece of data required on both parts of the structure? (seems like it's here at the top level and inside `FaultDomainConfig`?

Reply:
Yes, it's mandatory here, I checked the types of `FaultDomainConfig` in govmomi:
```
type VimClusterVSANStretchedClusterFaultDomainConfig struct {
	types.DynamicData

	FirstFdName   string                         `xml:"firstFdName"`
	FirstFdHosts  []types.ManagedObjectReference `xml:"firstFdHosts"`
	SecondFdName  string                         `xml:"secondFdName"`
	SecondFdHosts []types.ManagedObjectReference `xml:"secondFdHosts"`
}
```
It only stores info of two fault domains, but during enabling stretched cluster, we must choose one of them as preferred fault domain, that is the second fd could also be set as preferred fd.